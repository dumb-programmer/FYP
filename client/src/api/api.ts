export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
