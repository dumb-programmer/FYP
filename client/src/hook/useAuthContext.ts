import { AuthContext } from "@/context/AuthContext";
import { useContext } from "react";

export default function useAuthContext() {
  return useContext(AuthContext);
}
