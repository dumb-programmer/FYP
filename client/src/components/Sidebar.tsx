import useAuthContext from "@/hooks/useAuthContext";
import { useInfiniteQuery } from "react-query"
import { useNavigate } from "react-router-dom";
import CreateChatButton from "./CreateChatButton";
import ChatLink from "./ChatLink";
import { ArrowLeftStartOnRectangleIcon, Bars3Icon, ChevronDownIcon, ChevronLeftIcon, ChevronRightIcon } from "@heroicons/react/16/solid";
import { getChats, logout } from "@/api/api";
import { useLayoutEffect, useState } from "react";

export default function Sidebar() {
    const { data, refetch, hasNextPage, isFetchingNextPage, fetchNextPage } = useInfiniteQuery("chat-links", {
        queryFn: async ({ pageParam = 1 }) => {
            const response = await getChats(pageParam)
            if (response.ok) {
                return response.json();
            }
        },
        getNextPageParam: (lastPage) => {
            if (lastPage.hasMore) {
                return lastPage.nextPage
            }
        }
    });

    const { auth, setAuth } = useAuthContext();
    const navigate = useNavigate();
    const [show, setShow] = useState(false);

    useLayoutEffect(() => {
        const handleResize = () => {
            if (window.innerWidth >= 1024) {
                setShow(true);
            } else {
                setShow(false);
            }
        };

        window.addEventListener("resize", handleResize);

        handleResize();

        return () => window.removeEventListener("resize", handleResize);
    }, []);

    return <>
        {
            !show && <button className="btn btn-sm fixed top-6 left-2 z-10" onClick={() => setShow(show => !show)} title="open sidebar">
                <ChevronRightIcon height={18} width={18} />
            </button>
        }
        <aside className={`absolute lg:relative h-full z-10 menu prose bg-base-300 flex flex-col transition-all origin-left duration-150 ${!show ? "w-0 overflow-hidden bg-transparent" : "w-[300px] scale-x-100 after:content-[''] after:fixed after:inset-0 after:w-[calc(100vw-300px)] after:bg-black/50 after:left-[300px] after:-z-10 md:after:content-none"}`}>
            <button className="btn btn-sm absolute top-7 right-5 z-10" onClick={() => setShow(show => !show)} title="collapse sidebar">
                <ChevronLeftIcon height={18} width={18} />
            </button>
            <div className="flex flex-col justify-between h-full p-6">
                <div>
                    <h1 className="mt-10 text-3xl">Chats</h1>
                    <div className="pt-10 flex flex-col gap-6 overflow-auto h-5/6">
                        <CreateChatButton refetch={refetch} />
                        <nav className="flex flex-col overflow-y-auto h-1/2">
                            <ul>
                                {
                                    data && [].concat(...data.pages.map(page => page.chats)).map(chat => <ChatLink key={chat._id} chat={chat} refetch={refetch} />)
                                }
                            </ul>
                            {hasNextPage && <button data-testid="load-more-chats-btn" className="mt-2 btn btn-sm text-xs" onClick={() => fetchNextPage()} disabled={isFetchingNextPage}>Load More <ChevronDownIcon color="black" height={20} width={20} /> </button>}
                        </nav>
                    </div>
                </div>
                <div className="flex justify-between gap-3 items-center w-full">
                    {auth.profilePicture && <img className="h-12 w-12 rounded-full" src={auth.profilePicture} alt={`${auth.firstName} ${auth.lastName}`} />}
                    <div>
                        <p className="font-bold">{`${auth.firstName} ${auth.lastName}`}</p>
                        <span className="text-gray-600 text-sm">{auth?.email}</span>
                        <span className="text-gray-600">{auth?.username}</span>
                    </div>
                    <button title="logout" className="btn btn-ghost" onClick={async () => {
                        const response = await logout();
                        if (response.ok) {
                            setAuth(null);
                            navigate("/login");
                        }
                    }}>
                        <ArrowLeftStartOnRectangleIcon height={20} width={20} />
                    </button>
                </div>
            </div>
        </aside>
    </>
}