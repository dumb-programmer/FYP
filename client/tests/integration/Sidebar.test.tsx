import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import Sidebar from "../../src/components/Sidebar";
import { QueryClient, QueryClientProvider } from "react-query";
import { getChats, logout } from "../../src/api/api"
import { MemoryRouter, useNavigate } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import useAuthContext from "../../src/hooks/useAuthContext";

vi.mock("react-router-dom", async () => ({ ...await vi.importActual("react-router-dom"), useNavigate: vi.fn() }));
vi.mock("../../src/hooks/useAuthContext", () => ({ default: vi.fn() }));
vi.mock("../../src/api/api", async () => ({ ...await vi.importActual("../../src/api/api"), getChats: vi.fn(), logout: vi.fn() }))

const queryClient = new QueryClient();
const renderComponent = () => render(
    <MemoryRouter>
        <QueryClientProvider client={queryClient}>
            <Sidebar />
        </QueryClientProvider>
    </MemoryRouter>
);

describe("Sidebar", () => {
    const setAuth = vi.fn();
    const AuthObj = { firstName: "John", lastName: "Cena", email: "john@gmail.com" };
    const mockNavigate = vi.fn();

    beforeEach(() => {
        useAuthContext.mockReturnValue({ auth: AuthObj, setAuth })
        useNavigate.mockReturnValue(mockNavigate)
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("renders correctly", async () => {
        await waitFor(() => {
            renderComponent();
        })

        const component = screen.getByRole("complementary");

        expect(component).toMatchSnapshot();
    });

    it("getChats is called with the correct parameters the retrieved data is shown as chat links", async () => {
        const chats = [
            { _id: "1", name: "Chat 1" },
            { _id: "2", name: "Chat 2" },
        ];
        const response = {
            ok: true,
            json: async () => ({ chats, hasMore: false, nextPage: 2 }),
            hasMore: false
        }
        getChats.mockReturnValue(response)

        await waitFor(() => {
            renderComponent();
        })

        expect(getChats).toHaveBeenCalled();
        expect(getChats).toHaveBeenCalledWith(1);
        expect(screen.getByText(/chat 1/i)).toBeInTheDocument();
        expect(screen.getByText(/chat 2/i)).toBeInTheDocument();
        expect(screen.queryByRole("button", { name: /load more/i })).not.toBeInTheDocument();

    });

    it("load more button is shown, if there are more chat links to load", async () => {
        const chatsPage1 = [
            { _id: "1", name: "Chat 1" },
            { _id: "2", name: "Chat 2" },
        ];
        const chatsPage2 = [
            { _id: "3", name: "Chat 3" },
            { _id: "4", name: "Chat 4" },
        ];
        const response1 = {
            ok: true,
            json: async () => ({ chats: chatsPage1, hasMore: true, nextPage: 2 }),
        }
        const response2 = {
            ok: true,
            json: async () => ({ chats: chatsPage2, hasMore: false, nextPage: 3 }),
        }

        getChats.mockImplementation((page) => {
            if (page === 1) return response1;
            return response2;
        })

        await waitFor(() => {
            renderComponent();
        });


        expect(getChats).toHaveBeenCalled();
        expect(getChats).toHaveBeenCalledWith(1);
        expect(screen.getByText(/chat 1/i)).toBeInTheDocument();
        expect(screen.getByText(/chat 2/i)).toBeInTheDocument();
        expect(screen.queryByText(/chat 3/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/chat 4/i)).not.toBeInTheDocument();
        expect(screen.getByRole("button", { name: /load more/i })).toBeInTheDocument()


        await userEvent.click(screen.getByRole("button", { name: /load more/i }));

        expect(getChats).toHaveBeenCalled();
        expect(getChats).toHaveBeenCalledWith(2);
        expect(screen.getByText(/chat 1/i)).toBeInTheDocument();
        expect(screen.getByText(/chat 2/i)).toBeInTheDocument();
        expect(screen.getByText(/chat 3/i)).toBeInTheDocument();
        expect(screen.getByText(/chat 4/i)).toBeInTheDocument();

        // Since, there are no more chats to load, the load more button is hidden
        expect(screen.queryByRole("button", { name: /load more/i })).not.toBeInTheDocument()
    });

    it("logout button works", async () => {
        await waitFor(() => {
            renderComponent();
        })

        logout.mockReturnValue({ ok: true });

        await userEvent.click(screen.getByRole("button", { name: /logout/i }));

        expect(logout).toHaveBeenCalled();
        expect(setAuth).toHaveBeenCalled();

        // Navigate to /login
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith("/login");
    })
})