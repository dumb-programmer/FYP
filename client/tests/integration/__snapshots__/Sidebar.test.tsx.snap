// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > renders correctly 1`] = `
<aside
  class="menu prose relative p-10 bg-base-300 flex flex-col"
>
  <h1
    class="text-3xl"
  >
    Chats
  </h1>
  <div
    class="pt-10 flex flex-col gap-6 overflow-auto h-5/6"
  >
    <button
      class="btn ghost"
      data-testid="create-chat-btn"
    >
      <svg
        aria-hidden="true"
        data-slot="icon"
        fill="currentColor"
        height="20"
        viewBox="0 0 16 16"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.75 3.75a.75.75 0 0 0-1.5 0v3.5h-3.5a.75.75 0 0 0 0 1.5h3.5v3.5a.75.75 0 0 0 1.5 0v-3.5h3.5a.75.75 0 0 0 0-1.5h-3.5v-3.5Z"
        />
      </svg>
      Create Chat
    </button>
    <dialog
      class="modal"
      data-testid="modal"
      id=":r0:"
    >
      <div
        class="modal-box"
      >
        <h1
          class="text-2xl font-bold"
        >
          Create Chat
        </h1>
        <form
          class="flex flex-col gap-4 mt-4"
          id="create-chat-form"
        >
          <div
            class="form-control"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              class="input input-bordered"
              id="name"
              name="name"
              required=""
              type="text"
            />
          </div>
          <div
            class="form-control"
          >
            <label
              for="document"
            >
              Document
            </label>
            <input
              class="input file-input -ml-4 mt-2"
              id="document"
              name="document"
              required=""
              type="file"
            />
          </div>
        </form>
        <div
          class="modal-action"
        >
          <button
            class="btn btn-ghost"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            form="create-chat-form"
            type="submit"
          >
            Create
          </button>
        </div>
      </div>
    </dialog>
    <nav
      class="flex flex-col overflow-y-auto h-1/2"
    >
      <ul />
    </nav>
  </div>
  <div
    class="mt-auto flex gap-3 items-center"
  >
    <div>
      <p
        class="font-bold"
      >
        John Cena
      </p>
      <span
        class="text-gray-600 text-sm"
      >
        john@gmail.com
      </span>
      <span
        class="text-gray-600"
      />
    </div>
    <button
      class="btn btn-ghost"
      title="logout"
    >
      <svg
        aria-hidden="true"
        data-slot="icon"
        fill="currentColor"
        height="20"
        viewBox="0 0 16 16"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M14 4.75A2.75 2.75 0 0 0 11.25 2h-3A2.75 2.75 0 0 0 5.5 4.75v.5a.75.75 0 0 0 1.5 0v-.5c0-.69.56-1.25 1.25-1.25h3c.69 0 1.25.56 1.25 1.25v6.5c0 .69-.56 1.25-1.25 1.25h-3c-.69 0-1.25-.56-1.25-1.25v-.5a.75.75 0 0 0-1.5 0v.5A2.75 2.75 0 0 0 8.25 14h3A2.75 2.75 0 0 0 14 11.25v-6.5Zm-9.47.47a.75.75 0 0 0-1.06 0L1.22 7.47a.75.75 0 0 0 0 1.06l2.25 2.25a.75.75 0 1 0 1.06-1.06l-.97-.97h7.19a.75.75 0 0 0 0-1.5H3.56l.97-.97a.75.75 0 0 0 0-1.06Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</aside>
`;
